app/models/shared_ai_conversation.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+class SharedAiConversation < ActiveRecord::Base
+  DEFAULT_MAX_POSTS = 100
+
+  belongs_to :user
+  belongs_to :topic
+
+  validates :user_id, presence: true
+  validates :topic_id, presence: true
+  validates :share_key, presence: true, uniqueness: true
+
+  before_validation :generate_share_key, on: :create
+
+  def self.share_conversation(user, topic, max_posts: DEFAULT_MAX_POSTS)
+    conversation = find_by(user: user, topic: topic)
+    conversation_data = build_conversation_data(topic, max_posts: max_posts)
+
+    if conversation
+      conversation.update(**conversation_data)
+      conversation
+    else
+      create(user_id: user.id, topic_id: topic.id, **conversation_data)
+    end
+  end
+
+  class SharedPost
+    attr_accessor :user
+    attr_reader :id, :user_id, :created_at, :cooked
+    def initialize(post)
+      @id = post[:id]
+      @user_id = post[:user_id]
+      @created_at = DateTime.parse(post[:created_at])
+      @cooked = post[:cooked]
+    end
+  end
+
+  def populated_posts
+    return @populated_posts if @populated_posts
+    @populated_posts = posts.map { |post| SharedPost.new(post.symbolize_keys) }
```

Comment:
Since the params needed to share the post are quite small, might be better to just pass the ones you need:

```ruby
SharedPost.new(id: post.id, user_id: post.user_id, created_at: post.created_at, cooked: post.cooked)
```